src/toast/tod/crosstalk.py

```diff
@@ -0,0 +1,473 @@
+with the simple conventions that
+it has the following datasets:
+
+1. names: ASCII names of your detectors, say of length ``n``
+2. data: ``n`` by ``n`` float64 array of your crosstalk matrix ``m``, where you expect
+
+    ToD_crosstalked = m @ ToD_original
+
+    (Assuming row-major as is standard in Python.)
+
+Note that each crosstalk matrix is dense,
+and when multiple matrices are supplied, they are effectively block-diagonal
+(where each given dense matrix is a block.)
+I.e. in order to provide a block-diagonal crosstalk matrix,
+each block (together with the names of the detectors in that block)
+should write to a seperate HDF5 file to avoid unnessary computation.
+"""
+
+import argparse
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+import h5py
+import numpy as np
+
+try:
+    from numba import jit
+except ImportError:
+    jit = None
+
+from ..mpi import get_world
+from ..op import Operator
+from ..utils import Logger
+
+if TYPE_CHECKING:
+    from typing import List  # Optional
+
+    import toast
+
+# py37+
+# COMM: Optional[toast.mpi.Comm]
+# PROCS: int
+# RANK: int
+COMM, PROCS, RANK = get_world()
+logger = Logger.get()
+IS_SERIAL = PROCS == 1
+
+
+def _fma(out: 'np.ndarray[np.float64]', weights: 'np.ndarray[np.float64]', *arrays: 'np.ndarray[np.float64]'):
+    """Simple FMA, compiled to avoid Python memory implications.
+
+    :param out: must be zero array in the same shape of each array in `arrays`
+
+    If not compiled, a lot of Python objects will be created,
+    and as the Python garbage collector is inefficient,
+    it would have larger memory footprints.
+    """
+    for weight, array in zip(weights, arrays):
+        out += weight * array
+
+
+if jit is None:
+    logger.warning('Numba not present. _fma in crosstalk will have more intermediate Numpy array objects created that uses more memory.')
+else:
+    # cache is False to avoid IO on HPC.
+    _fma = jit(_fma, nopython=True, nogil=True, cache=False)
+
+
+def add_crosstalk_args(parser: 'argparse.ArgumentParser'):
+    """Add crosstalk args to argparse.ArgumentParser object."""
+    parser.add_argument(
+        "--crosstalk-matrix",
+        type=Path,
+        nargs='*',
+        required=False,
+        help="input path(s) to crosstalk matrix in HDF5 container.",
+    )
+    parser.add_argument(
+        "--crosstalk-matrix-debug",
+        type=bool,
+        action='store_true',
+        required=False,
+        help="if specified, perform more checks and emit more messages. You may want to set TOAST_LOGLEVEL=DEBUG as well.",
+    )
+
+
+class SimpleCrosstalkMatrix:
+    """A thin crosstalk matrix class.
+
+    This is a simple container storing the crosstalk matrix,
+    but not generating it.
+
+    The length of `names` should match the dimension of `data`, which should be a square array.
+    Runtime checking is done only if `debug` is set to `True`.
+    """
+
+    def __init__(
+        self,
+        names: "np.ndarray['S']",
```

Comment:
It is because the typical scenario would be reading them from HDF5 files, would already provide `np.ndarray['S']` type. So I figure the internal data structure should be as closely to those from HDF5 as possible.

But the end user creating them certainly would be more troublesome. May be provide a classmethod?

Reply:
Resolved in 6969304c. I didn't use classmethod as there's a better way.